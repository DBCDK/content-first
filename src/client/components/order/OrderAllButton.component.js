import React from 'react';
import Button from '../base/Button';
import T from '../base/T';
import withPermisisons, {ORDER_CONTEXT} from '../hoc/Permissions';

const OrderAllButton = props => {
  return '';
  /*(
  <Button
    size="medium"
    type="quaternary"
    className="orderAllBtn"
    data-cy="orderAllButton"
    {...props}
  >
    <T component="shortlist" name="shortlistOrder" />
  </Button>
  )*/};

export default withPermisisons(OrderAllButton, ORDER_CONTEXT);
