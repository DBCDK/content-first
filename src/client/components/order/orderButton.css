.orderButton.progress {
  display: flex;
  height: auto;
  align-items: center;
}

.orderButton.success {
  background-color: var(--pistache) !important;
}

.orderButton.error {
  background-color: var(--fersken) !important;
}
