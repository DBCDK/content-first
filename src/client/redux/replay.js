import {setItem, getItem} from '../utils/localstorage';

const REPLAY_BEGIN = 'REPLAY_BEGIN';
const REPLAY_END = 'REPLAY_END';

const LOCAL_STORAGE_REPLAY_KEY = 'contentfirstreplay';
const LOCAL_STORAGE_REPLAY_VERSION = 1;

const REPLAY_PATH = '/replay';

// Put actions here, which should be listened for
// and put in replay queue
const replayableActions = {
  ORDER: action => action,
  ON_LOCATION_CHANGE: action =>
    action.path !== REPLAY_PATH
      ? {type: 'HISTORY_REPLACE', path: action.path + action.location.search}
      : null
};

// Some times we want to clear the replay queue,
// for instance, at page changes and when modal closes
const clearActions = {
  CLOSE_MODAL: () => true,
  ON_LOCATION_CHANGE: action => action.path !== REPLAY_PATH
};

// Actions which will never be deleted from replay queue
// For isntance, we always want to replay the latest location
const keepAfterClear = {
  ON_LOCATION_CHANGE: true
};

// Actions which will be replaced. I.e. Multiple entries not allowed
const replaceExisting = {
  ON_LOCATION_CHANGE: true
};

// Actions which will trigger dispatch of actions in replay queue
const replayBeginWhen = {
  ON_USER_DETAILS_ERROR: state => state.routerReducer.path === REPLAY_PATH,
  ON_USER_DETAILS_RESPONSE: state => state.routerReducer.path === REPLAY_PATH
};

export const replayReducer = (
  state = getItem(LOCAL_STORAGE_REPLAY_KEY, LOCAL_STORAGE_REPLAY_VERSION, {
    isReplaying: false,
    actions: []
  }),
  action
) => {
  if (action.type === REPLAY_BEGIN) {
    return {...state, isReplaying: true};
  }
  if (action.type === REPLAY_END) {
    return {
      ...state,
      isReplaying: false,
      actions: state.actions.filter(a => keepAfterClear[a.type])
    };
  }
  if (state.isReplaying) {
    return state;
  }
  if (clearActions[action.type] && clearActions[action.type](action)) {
    state = {
      ...state,
      actions: state.actions.filter(a => keepAfterClear[a.type])
    };
  }
  if (
    replayableActions[action.type] &&
    replayableActions[action.type](action)
  ) {
    const actions = replaceExisting[action.type]
      ? state.actions.filter(a => a.type !== action.type)
      : state.actions;
    state = {...state, actions: [...actions, action]};
  }
  return state;
};

export const replayMiddleware = store => next => action => {
  const prevReplay = store.getState().replay;
  const res = next(action);
  const {replay} = store.getState();

  if (
    replayBeginWhen[action.type] &&
    replayBeginWhen[action.type](store.getState())
  ) {
    store.dispatch({type: REPLAY_BEGIN});
    replay.actions.forEach(a => store.dispatch(replayableActions[a.type](a)));
    store.dispatch({type: REPLAY_END});
  }

  if (prevReplay !== replay) {
    setItem(LOCAL_STORAGE_REPLAY_KEY, replay, LOCAL_STORAGE_REPLAY_VERSION);
  }
  return res;
};
