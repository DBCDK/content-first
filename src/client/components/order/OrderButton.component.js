import React from 'react';
import Button from '../base/Button';
import Icon from '../base/Icon';
import T from '../base/T';
import Spinner from '../general/Spinner/Spinner.component';
import {withWork} from '../hoc/Work';
import withPermisisons from '../hoc/Permissions';
import Kiosk from '../base/Kiosk/Kiosk';

import './orderButton.css';
import FindBookButton from '../kiosk/WayFinder/FindBookButton';

export function OrderButton(props) {
  // OrderButton default state:

  const orderState = (state => {
    switch (state) {
      // Ordered button state
      case 'ordered':
        return {
          class: 'success',
          label: <T component="order" name="orderDone" />
        };

      // Order in progres
      case 'ordering':
        return {
          class: 'progress',
          label: (
            <React.Fragment>
              <Spinner className="mr-1" size={14} color={'var(--petroleum)'} />
              <T component="order" name="orderInProgress" />
            </React.Fragment>
          )
        };

      // Error while ordering
      case 'error':
        return {
          class: 'progress',
          label: <T component="order" name="orderError" />
        };

      // default order state
      default:
        return {
          class: '',
          label: (
            <React.Fragment>
              <Icon name={props.icon} />
              {props.children || props.label}
            </React.Fragment>
          )
        };
    }
  })(props.orderState);

  // If physical book does NOT exist in the work collection
  if (!props.collectionContainsBook()) {
    return null;
  }

  return (
    <Kiosk
      render={({kiosk}) => {
        if (kiosk.enabled) {
          return <FindBookButton pid={props.pid} />;
        }
        return (
          <Button
            className={`orderButton ${orderState.class} ${props.className}`}
            type={props.type}
            size={props.size}
            iconLeft={props.iconLeft}
            iconRight={props.iconRight}
            onClick={props.onClick || props.order}
            dataCy="order-btn"
          >
            {orderState.label}
          </Button>
        );
      }}
    />
  );
}

export default withPermisisons(
  withWork(OrderButton, {
    includeCollection: true
  }),
  {
    name: 'OrderButton',
    modals: {
      login: {
        context: {
          title: 'Bestil bøger'
        }
      },
      premium: {
        context: {
          title: 'Bestil bøger',
          reason: 'Bestilling af bøger er ikke tilgængeligt for dit bibliotek'
        }
      }
    }
  }
);
