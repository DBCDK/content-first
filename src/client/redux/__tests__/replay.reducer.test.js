import {replayReducer} from '../replay';
import {ORDER} from '../order.reducer';
import {ON_LOCATION_CHANGE} from '../router.reducer';

const createTestState = () => ({
  expanded: false,
  elements: [],
  isLoading: false,
  pendingMerge: null
});

const initState = {
  isReplaying: false,
  actions: []
};

describe('replayReducer', () => {
  test('Allowed actions are recorded', () => {
    let resState = replayReducer(initState, {
      type: ON_LOCATION_CHANGE,
      location: {search: '?param=val'}
    });
    resState = replayReducer(resState, {type: ORDER, somekey: 1});
    resState = replayReducer(resState, {type: ORDER, somekey: 2});
    expect(resState).toMatchSnapshot();
  });
  test('Actions are not recorded if they are not allowed', () => {
    let resState = replayReducer(initState, {type: 'NOT_ALLOWED'});
    expect(resState).toMatchSnapshot();
  });
  test('ON_LOCATION_CHANGE replaces existing action of this type', () => {
    let resState = replayReducer(initState, {
      type: ON_LOCATION_CHANGE,
      location: {search: '?param=val'}
    });
    resState = replayReducer(initState, {
      type: ON_LOCATION_CHANGE,
      location: {search: '?param=val2'}
    });
    expect(resState).toMatchSnapshot();
  });
  test('ON_LOCATION_CHANGE clears actions', () => {
    let resState = replayReducer(initState, {
      type: ON_LOCATION_CHANGE,
      location: {search: '?param=val'}
    });
    resState = replayReducer(resState, {type: ORDER, somekey: 1});
    resState = replayReducer(resState, {type: ORDER, somekey: 2});
    resState = replayReducer(initState, {
      type: ON_LOCATION_CHANGE,
      location: {search: '?param=val2'}
    });
    expect(resState).toMatchSnapshot();
  });
});
