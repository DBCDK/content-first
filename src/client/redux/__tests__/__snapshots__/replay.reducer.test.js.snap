// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replayReducer Actions are not recorded if they are not allowed 1`] = `
Object {
  "actions": Array [],
  "isReplaying": false,
}
`;

exports[`replayReducer Allowed actions are recorded 1`] = `
Object {
  "actions": Array [
    Object {
      "location": Object {
        "search": "?param=val",
      },
      "type": "ON_LOCATION_CHANGE",
    },
    Object {
      "somekey": 1,
      "type": "ORDER",
    },
    Object {
      "somekey": 2,
      "type": "ORDER",
    },
  ],
  "isReplaying": false,
}
`;

exports[`replayReducer ON_LOCATION_CHANGE clears actions 1`] = `
Object {
  "actions": Array [
    Object {
      "location": Object {
        "search": "?param=val2",
      },
      "type": "ON_LOCATION_CHANGE",
    },
  ],
  "isReplaying": false,
}
`;

exports[`replayReducer ON_LOCATION_CHANGE replaces existing action of this type 1`] = `
Object {
  "actions": Array [
    Object {
      "location": Object {
        "search": "?param=val2",
      },
      "type": "ON_LOCATION_CHANGE",
    },
  ],
  "isReplaying": false,
}
`;
